File: eon_timer/timers/calibrator.py
import math
from typing import Final

from eon_timer.settings.timer.model import Console, TimerSettingsModel


class Calibrator:
    def __init__(self, timer_settings: TimerSettingsModel) -> None:
        self.timer_settings: Final = timer_settings

    def to_delays(self, milliseconds: float) -> int:
        return math.floor(milliseconds / self.framerate)

    def to_milliseconds(self, delays: int) -> float:
        return self.framerate * delays

    def calibrate_to_delays(self, milliseconds: float) -> int:
        return round(milliseconds) if self.precision_calibration else self.to_delays(milliseconds)

    def calibrate_to_milliseconds(self, delays: int) -> float:
        return float(delays) if self.precision_calibration else self.to_milliseconds(delays)

    def create_calibration(self, delays: int, seconds: float) -> float:
        return self.to_milliseconds(delays - self.to_delays(seconds * 1000))

    @property
    def console(self) -> Console:
        return self.timer_settings.console.get()

    @property
    def precision_calibration(self) -> bool:
        return self.timer_settings.precision_calibration.get()

    @property
    def minimum_length(self) -> int:
        return self.timer_settings.minimum_length.get() * 1000

    @property
    def framerate(self) -> float:
        custom_framerate = self.timer_settings.custom_framerate.get()
        if custom_framerate == 0:
            raise ValueError('Custom framerate must be greater than 0')
        return self.console.framerate or (1000 / custom_framerate)

File: src/timers/calibrator.ts
import {
  GBA_MS_PER_FRAME,
  NDS_SLOT1_MS_PER_FRAME,
  NDS_SLOT2_MS_PER_FRAME,
} from '../utils/constants';
import { Console } from '../utils/types';

export interface CalibratorSettings {
  console: Console;
  customFramerate: number;
  precisionCalibration: boolean;
  minimumLength: number; // in milliseconds
}

// Match the desktop timer's C# Math.Round(decimal) midpoint-to-even behavior.
function roundHalfToEven(value: number): number {
  if (!Number.isFinite(value)) {
    return Math.round(value);
  }

  const lower = Math.floor(value);
  const upper = Math.ceil(value);
  if (lower === upper) {
    return lower;
  }

  const lowerDistance = value - lower;
  const upperDistance = upper - value;
  const epsilon = Number.EPSILON * Math.max(1, Math.abs(value));

  if (Math.abs(lowerDistance - upperDistance) <= epsilon) {
    return Math.abs(lower) % 2 === 0 ? lower : upper;
  }

  return lowerDistance < upperDistance ? lower : upper;
}

function getMsPerFrame(settings: CalibratorSettings): number {
  switch (settings.console) {
    case Console.GBA:
      return GBA_MS_PER_FRAME;
    case Console.NDS_SLOT2:
      return NDS_SLOT2_MS_PER_FRAME;
    case Console.NDS_SLOT1:
    case Console.DSI:
    case Console.THREE_DS:
      return NDS_SLOT1_MS_PER_FRAME;
    case Console.CUSTOM: {
      if (settings.customFramerate === 0) {
        throw new Error('Custom framerate must be greater than 0');
      }
      return 1000 / settings.customFramerate;
    }
    default:
      return NDS_SLOT1_MS_PER_FRAME;
  }
}

export function toDelays(settings: CalibratorSettings, milliseconds: number): number {
  return roundHalfToEven(milliseconds / getMsPerFrame(settings));
}

export function toMilliseconds(settings: CalibratorSettings, delays: number): number {
  return roundHalfToEven(getMsPerFrame(settings) * delays);
}

export function calibrateToDelays(settings: CalibratorSettings, milliseconds: number): number {
  return settings.precisionCalibration
    ? roundHalfToEven(milliseconds)
    : toDelays(settings, milliseconds);
}

export function calibrateToMilliseconds(settings: CalibratorSettings, delays: number): number {
  return settings.precisionCalibration ? delays : toMilliseconds(settings, delays);
}

export function createCalibration(
  settings: CalibratorSettings,
  delays: number,
  seconds: number,
): number {
  return toMilliseconds(settings, delays - toDelays(settings, seconds * 1000));
}
